pkg/autoupdate/tuf/library_manager.go

```diff
@@ -320,53 +246,246 @@ func (ulm *updateLibraryManager) tidyUpdateLibrary(binary autoupdatableBinary, c
 			continue
 		}
 
-		// Only keep good executables
-		versionDir := filepath.Join(ulm.updatesDirectory(binary), versionsInLibrary[i].Original())
+		versionDir := filepath.Join(ulm.updatesDirectory(binary), rawVersion)
 		if err := autoupdate.CheckExecutable(context.TODO(), executableLocation(versionDir, binary), "--version"); err != nil {
-			ulm.removeUpdate(binary, versionsInLibrary[i].Original())
+			invalidVersions = append(invalidVersions, rawVersion)
 			continue
 		}
 
-		nonCurrentlyRunningVersionsKept += 1
+		versionsInLibrary = append(versionsInLibrary, v)
+	}
+
+	// Sort the versions (ascending order)
+	sort.Sort(semver.Collection(versionsInLibrary))
+
+	// Transform versions back into strings now that we've finished sorting them
+	versionsInLibraryStr := make([]string, len(versionsInLibrary))
+	for i, v := range versionsInLibrary {
+		versionsInLibraryStr[i] = v.Original()
 	}
+
+	return versionsInLibraryStr, invalidVersions, nil
+}
+
+// IsInstallVersion checks whether the version in the given target is the same as the one
+// contained in the installation.
+func (ulm *updateLibraryManager) IsInstallVersion(binary autoupdatableBinary, targetFilename string) bool {
+	installedVersion, _, err := ulm.installedVersion(binary)
+	if err != nil {
+		level.Debug(ulm.logger).Log("msg", "could not get installed version", "binary", binary, "err", err)
+		return false
+	}
+
+	return installedVersion.Original() == versionFromTarget(binary, targetFilename)
+}
+
+// installedVersion returns the version of, and the path to, the originally-installed binary.
+func (ulm *updateLibraryManager) installedVersion(binary autoupdatableBinary) (*semver.Version, string, error) {
+	pathToBinary := ulm.findInstallLocation(binary)
+	if pathToBinary == "" {
+		return nil, "", fmt.Errorf("could not find install location for `%s`", binary)
+	}
+
+	if cachedVersion, err := ulm.getCachedInstalledVersion(binary); err == nil {
+		return cachedVersion, pathToBinary, nil
+	}
+
+	cmd := exec.Command(pathToBinary, "--version")
+	cmd.Env = append(cmd.Env, "LAUNCHER_SKIP_UPDATES=TRUE") // Prevents launcher from fork-bombing
+	out, err := cmd.Output()
+	if err != nil {
+		return nil, "", fmt.Errorf("could not execute %s --version: out %s, error %w", pathToBinary, string(out), err)
+	}
+
+	var v *semver.Version
+	switch binary {
+	case binaryLauncher:
+		v, err = parseLauncherVersion(out)
+	case binaryOsqueryd:
+		v, err = parseOsquerydVersion(out)
+	default:
+		return nil, "", fmt.Errorf("cannot parse version for unknown binary %s", binary)
+	}
+
+	if err != nil {
+		return nil, "", fmt.Errorf("could not parse binary install version: %w", err)
+	}
+
+	ulm.cacheInstalledVersion(binary, v)
+
+	return v, pathToBinary, nil
+}
+
+// findInstallLocation attempts to find the install location for the given binary, looking
+// in well-known locations.
+func (ulm *updateLibraryManager) findInstallLocation(binary autoupdatableBinary) string {
```

Comment:
This, especially, feels like it's going to be fragile and hard to manage.

If we really want to track the installed version, I think we should set an ENV on launch and carry it through.

Reply:
I think you're right -- it's already felt pretty fragile in development. Maybe it's fine to just not care about it -- I'm fine with setting it on launch and carrying it through, too.